boltrouter/source_bucket.go

```diff
@@ -0,0 +1,84 @@
+package boltrouter
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+type s3RequestStyle string
+
+const (
+	virtualHostedStyle s3RequestStyle = "virtual-hosted-style"
+	pathStyle          s3RequestStyle = "path-style"
+	nAuthDummy         s3RequestStyle = "n-auth-dummy"
+)
+
+type SourceBucket struct {
+	Bucket string
+	Region string
+	Style  s3RequestStyle
+}
+
+// extractSourceBucket extracts the aws request bucket using Path-style or Virtual-hosted-style requests.
+// https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html
+// This method will "n-auth-dummy" if nothing is found
+func extractSourceBucket(ctx context.Context, req *http.Request) (SourceBucket, error) {
+	region, err := getRegionForBucket(ctx, req.Header.Get("Authorization"))
+	if err != nil {
+		return SourceBucket{}, fmt.Errorf("could not get region for bucket: %w", err)
+	}
+
+	ret := SourceBucket{
+		Region: region,
+		Bucket: "n-auth-dummy",
+		Style:  nAuthDummy,
+	}
+
+	isVirtualHostedStyle := false
+	split := strings.Split(req.Host, ".")
+	if len(split) > 1 {
+		if _, err := strconv.Atoi(split[0]); err != nil {
+			// is not a number, so it is a bucket name
+			isVirtualHostedStyle = true
```

Comment:
pathstyle host looks like this `Host: s3.us-east-1.amazonaws.com`
virtual hosted host looks like this `Host: DOC-EXAMPLE-BUCKET1.eu.s3.eu-west-1.amazonaws.com`

so split[0] yields `s3` and `DOC-EXAMPLE-BUCKET1` respectively?